### PR TITLE
feat: cli init

### LINKED ISSUE #95: Command-Line Tool Totaling Contributor Rewards for UBQ Airdrop
## Overview

We are looking to develop a command-line tool that automates the tallying of contributions across all Ubiquity repositories for the last year. This tool will play a crucial role in recognizing and rewarding our contributors' efforts through UBQ token payouts.

### Tally Contributions

The tool should accurately parse through all the "permits" (or payments for recognized contributions) generated across our repositories to identify and sum up each contributor's total contributions in terms of dollars earned.

### UBQ Token Calculation

Based on the dollar value of each contributor's total contributions, calculate the equivalent amount in UBQ tokens, with the conversion rate being 1 UBQ per dollar earned.

### Output Format

The results should be compiled into a comprehensive CSV file for each repository, detailing the contributors' names, total dollar earnings, and corresponding UBQ tokens. This CSV file will serve as the basis for UBQ token payouts.

### Automation and Execution
This script should be designed for easy execution at specific times, particularly when enabling UBQ token payouts to ensure a seamless distribution process.

## Requirements
- Use TypeScript for development to ensure type safety and maintainability.
- Utilize the @octokit/rest library for GitHub API interactions to fetch permits across repositories.
- Ensure the tool is executable through `tsx` to facilitate integration into our existing workflow.
- Include a testing mechanism that prints test results, which will be a critical part of the pull request QA process.

## Final Deliverables
- A fully functional command-line tool meeting the above specifications.
- Documentation on how to set up, use, and contribute to the tool.
- Test cases and results demonstrating the tool's effectiveness and accuracy.
- This tool is integral to our efforts to automate and streamline our reward system for contributors. It will not only enhance our operational efficiency but also ensure transparency and fairness in UBQ token distribution.

We look forward to seeing innovative solutions and contributions from the community to bring this project to fruition.

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #95 

Comment based parsing as opposed to permit based.

Accounting for the multiple formats going back as far as the start of 2023 I never went back any further thinking it would only be covering the last year.

See readme for usage instructions.

Originally I started with merged pull requests then grabbed the closing issue reference and worked backwards that way but after trying it from an all issues approach it felt like the results were better.

> No we should include every contribution that exists.

So it's no longer just tallying contributions from the last year as per the spec but all time payments?

## QA
- See generated files in root
- Action run: https://github.com/Keyrxng/ubq-airdrop-cli/actions/runs/7860439292

### CODE REVIEW

File: airdrop-cli/all_repos_contributors.csv
Comment (pavlovcik) [inline airdrop-cli/all_repos_contributors.csv:1]: You should add `.csv` to `.gitignore`!

Comment (Keyrxng) [reply airdrop-cli/all_repos_contributors.csv:1]: Will do, I thought it would be relevant for QA so left them in

File: airdrop-cli/webpack.config.js
Comment (pavlovcik) [inline airdrop-cli/webpack.config.js:1]: Why webpack? `tsx` should let you execute the typescript files directly, there should not be any compilation.

File: airdrop-cli/package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  "scripts": {
+    "start": "node dist/index.js",
+    "tally": "node dist/index.js tally",
+    "tally-multi-csv": "node dist/index.js tally-multi-csv",
+    "tally-from": "node dist/index.js tally-from",
+    "single": "node dist/index.js single",
+    "help": "node dist/index.js help",
+    "build": "npx tsc",
+    "test": "npx jest"
+  },
```
Comment (pavlovcik) [inline airdrop-cli/package.json:16]: Looks like I forgot to include as part of the spec to use https://github.com/ubiquity/ts-template

It sets up a lot of the project for you, but also includes some essential linting rules that will make your code higher quality. The easiest way to include it in your work now is to 
```sh
git merge --allow-unrelated-histories
```

File: airdrop-cli/all_repos_contributors.csv
Comment (pavlovcik) [reply airdrop-cli/all_repos_contributors.csv:1]: Makes sense. I'll look at the CSVs when I spot check later.

File: airdrop-cli/webpack.config.js
Comment (Keyrxng) [reply airdrop-cli/webpack.config.js:1]: copy/paste from one of the other CLI's I've built, I'll use the ts-template as you suggest below

File: airdrop-cli/all_repos_contributors.csv
Comment (Keyrxng) [reply airdrop-cli/all_repos_contributors.csv:1]: I've removed them now bringing over the ts-template but you can get them from the action files for spot checking although it's only the ubiquibot repo. If you want them I'll drop them back in otherwise you could run the cli yourself.

As I'm unfamiliar with the template I didn't remove any scaffolding although it does look like quite a bit could be removed just wasn't sure

File: airdrop-cli/src/commands/tally-multi-csv.ts [hunk 2]
```diff
@@ -3,14 +3,14 @@ import { fetchPublicRepositories, processRepo } from "../invoke";
 import { loadingBar, writeCSV } from "../utils";
 
 @command({
```
Comment (pavlovcik) [inline airdrop-cli/src/commands/tally-multi-csv.ts:5]: I had no idea that `@` is valid syntax. This code is interesting!

File: airdrop-cli/src/commands/tally.ts [hunk 3]
```diff
@@ -6,8 +6,7 @@ import { invoke } from "../invoke";
 
 @command({
   brief: "Tally UBQ airdrop.",
-  description:
-    "Includes all repositories with payments on dev branch since 01-01-2023.",
+  description: "Includes all repositories with payments on dev branch since 01-01-2023.",
```
Comment (pavlovcik) [inline airdrop-cli/src/commands/tally.ts:9]: What does "on dev branch" mean? Issues are not associated with branches.

File: airdrop-cli/src/commands/tally-multi-csv.ts (hunk 2 above)
Comment (Keyrxng) [reply airdrop-cli/src/commands/tally-multi-csv.ts:5]: It's a decorator, often times they throw errors depending on build settings. Pretty sure they are an older thing you rarely see them anywhere anywhere

**Review by pavlovcik — COMMENTED** [review]
Code is looking pretty solid. Still have to get around to do that spot check but I'm optimistic.

File: airdrop-cli/src/commands/tally.ts (hunk 3 above)
Comment (Keyrxng) [reply airdrop-cli/src/commands/tally.ts:9]: well it's actually the repo default branch which is typically development for UBQ repos but it's whatever the current default branch

**Comment by Keyrxng** [comment]
After you give this a look over let me know if this is the way forward of if the permit based approach would be better?

I think this with a bit more refinement would be the more comprehensive approach

I started a fresh bot with fresh DB and I do not have any permits. So when you said just parse the permits you meant scrape the claim urls from comments and decode the permit data from that?

**Comment by pavlovcik** [comment]
I think that parsing the encoded permits from within the GitHub comments is fine. I consider basically all of those already transferred and credited for. I only invalidated a few, and its okay to give credit for them for now.

**Comment by Keyrxng** [comment]
validating the non_payment repos I found one newer style comment that was missed so I'll account for that.

I'll try improve the debugging data as well

**Comment by pavlovcik** [comment]
I appreciate you taking the time to check your work. You can mark this as ready to review when you resolve that.

**Comment by Keyrxng** [comment]
- Now pulling all permits: 448 tracked
- Permits assigned to users: 434
- Only format unaccounted for is: 
``0xcodercrane **[ [ CLAIM 500 ]** ]()for 0x00868BB3BA2B36316c2fc42E4aFB6D4246b77E46`` but only seen [here](https://github.com/Ubiquity/ubiquibot/issues/165) which looks to be old testing formats.

- Root outputs: User UBQ Totals, All Payments, All Permits, All Repos w/o Payment

I've caught every case I can find and done my best to attribute the right payment to the correct user with the correct reward origin

Validating using my own figures as before it appears to be much better

There are about 25ish attributions to "No Assignee" but can be verified easily enough, I'll be happy to do that once we know the data processing logic isn't changing again

<details>
<summary>Tally Table</summary>

```
devpanther,5335.849999999999
gitcoindev,4314.2
No assignee,4249.799999999999
seprintour,3833.95
wannacfuture,3750.35
sergfeldman,3743.7
Steveantor,3641.45
molecula451,2917.8
0xcodercrane,2284.4500000000003
Hodlatoor,1832.0000000000002
whilefoo,1774.9499999999998
Keyrxng,1574.5
byteballet,1336.8
pavlovcik,1294.0000000000002
zgorizzo69,1200
rndquu,1097.1999999999998
Venoox,875
bojan07,871.0000000000001
eugenioclrc,850
web4er,806.25
EtherealGlow,761.35
BeanieMen,639.4
FibrinLab,300
Draeieg,275.65
Sadaf-A,214.14999999999998
diamondnegroni,202.7
harjaapdhillon16,200
PhantomCracker,184.89999999999998
baibhavKumar1,124.5
pwilson77,100
sweetapplepumpkinpie,56.25
b4s36t4,56.25
devtestubq,56.25
rotcivegaf,50
starlitnightsky,37.5
tungbq,25
aditygrg2,18.75
rndquu2,16
```

</details>

**Comment by Keyrxng** [comment]
Is this admin wallet address? As in the primary payment address for UBQ

0xf87ca4583C792212e52720d127E7E0A38B818aD1

If there are more could you give me them please?

**Comment by Keyrxng** [comment]
I don't think I can take this any further without input on it as is, naturally there will be some manual validation required regardless but if you've any ideas on how to further improve the validity lmk

**Comment by pavlovcik** [comment]
Hey last thing but I'm hoping to see a workflow run can you add that so I can have it run in the GitHub action? 

You just need to yarn add tsx and then yarn tsx the script name and it should work.

**Comment by Keyrxng** [comment]
Is tsx a requirement if it's running fine with tsc? I can run the action I used before and output the files to the action logs but that doesn't use tsx.

The trouble is that tsx accepts some of the tsconfig props but not all, takes no custom config itself and throws silly type errors for packages specifically viem's utility types - (info from here)[https://www.npmjs.com/package/tsx]

If it's a requirement I'll see about a workaround

**Comment by Keyrxng** [comment]
I'll see what I can do with it then pav and I'll make the final push in an hour or so, should be a bit cleaner for you to review too

**Comment by Keyrxng** [comment]
- Action run: https://github.com/Keyrxng/ubq-airdrop-cli/actions/runs/7890222633

Good to go now, the gha will be deleted I assume as it's only for QA?

also the commitlint seems to be failing on the ts-template, I tried to solve it using the ubq-dollar repo config file and the example config they provide through the error, swapping out the package.json mention of it but it just kept repeatedly failing for me telling me the rules were empty and to set some.

Only noticed it after trying to commit while inside my original airdrop-cli repo after having brought in the template whereas before I had been working from this repo inside /airdrop-cli so none of the hooks were hitting

**Comment by pavlovcik** [comment]
I have to see the commit lint error you're talking about because I haven't experienced that. 

I think we can keep the GitHub Action. 

I started spot checking the CSVs and so far so good. I was a bit confused about the 25 permits that failed to parse in the CI log. There also seems to be a mismatch with the total amounts of payment permits issued and how many are on chain (it was like 350 on chain and like 450 in the logs if I remember correctly) I had to suddenly get off my computer am and posting this comment from a taxi.

**Comment by Keyrxng** [comment]
Appreciate it Pav, I'm hoping to get this merged asap bit tight for cash atm lmao

I'll give things another once over later and see if I can improve the output any further. It's been a balancing act trying to capture everything and create useful datasets without swarming them with noise while collecting actionable debug data.

Once the CLI is complete and ready for use before the airdrop, is your vision for it to remove the need for manual validation all together or will results still be validated manually considering?

**Comment by pavlovcik** [comment]
Can give the full review post merge given the delay and initial approval

File: .github/workflows/tallyRepo.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v3
+
+      - name: Install Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: "20"
```
Comment (pavlovcik) [inline .github/workflows/tallyRepo.yml:22]:
```suggestion
          node-version: "20.10.0"
```

File: airdrop-cli/.github/workflows/tallyRepo.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v3
+
+      - name: Install Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: "20"
```
Comment (pavlovcik) [inline airdrop-cli/.github/workflows/tallyRepo.yml:21]:
```suggestion
          node-version: "20.10.0"
```

File: airdrop-cli/src/commands/debug.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Command, command, metadata } from "clime";
+import { loadingBar } from "../utils";
+import { parseDebugData } from "../utils/debug";
+
+@command({
+  brief: "Debug CLI data.",
+  description: "Displays and processes the the available debug data.",
+})
+export default class extends Command {
+  @metadata
```
Comment (pavlovcik) [inline airdrop-cli/src/commands/debug.ts:10]: 👀

File: airdrop-cli/src/notes.md
Comment (pavlovcik) [inline airdrop-cli/src/notes.md:1]: Is this file necessary?

**Review by pavlovcik — APPROVED** [review]

**Comment by pavlovcik** [comment]
Screenshot 2024-02-15 at 03 01 15
lol i need to change the settings

File: airdrop-cli/src/commands/debug.ts (hunk 6 above)
Comment (Keyrxng) [reply airdrop-cli/src/commands/debug.ts:10]: I'll be sure to remove those

File: airdrop-cli/src/notes.md
Comment (Keyrxng) [reply airdrop-cli/src/notes.md:1]: it is not I was using it to track a few things I'll be sure to remove it

**Comment by Keyrxng** [comment]
Appreciate it Pav thank you

**Comment by pavlovcik** [comment]
Why is it DEBUG and not parsing the correct amount? Looks like there are mistakes.

Issue # | Amount | Repository | Currency | Payee | Type | URL |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
14 | 200 | devpool-directory-bounties | WXDAI | FernandVEYRIER | assignee | https://github.com/Ubiquity/devpool-directory-bounties/issues/14 |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |  
12 | 0 | work.ubq.fi | DEBUG | FernandVEYRIER | assignee | https://github.com/Ubiquity/work.ubq.fi/issues/12 |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |

**Comment by pavlovcik** [comment]
I realize now that I asked for addresses to be the identifiers instead of the user name. This is relevant because I jumbled up the addresses in the database at some point. Why dont you change the script to use addresses? Like I am pretty sure I mentioned in the specification, for the airdrop I only need addresses (GitHub username doesn't help me send an airdrop)

5 | 0 | ubiquibot-kernel | DEBUG | whilefoo | conversation | https://github.com/Ubiquity/ubiquibot-kernel/issues/5 | FALSE
-- | -- | -- | -- | -- | -- | -- | --

**Comment by pavlovcik** [comment]
@Keyrxng I just finished a high level spot check. The only thing that really needs to be changed is to use addresses instead I think.
